platform-hub-api/app/models/kubernetes_token.rb

```diff
@@ -1,18 +1,142 @@
-class KubernetesToken < KubernetesTokenBase
-  attr_accessor :identity_id, :expire_privileged_at
+class KubernetesToken < ApplicationRecord
+  TOKEN_LENGTH = 36
+  UID_LENGTH = 36
+  PRIVILEGED_GROUP_MAX_EXPIRATION_SECONDS = 21600 # max privilege period is 6h
 
-  validates_presence_of :identity_id
+  NAME_REGEX = /\A[a-zA-Z][\@\.\w-]*\z/
 
-  belongs_to :identity
+  include Audited
+  audited descriptor_field: :name, associated_field: :tokenable
 
-  def self.from_data(attributes)
-    attributes = attributes.with_indifferent_access
-    params = attributes
-      .extract!(:identity_id, :cluster, :uid, :groups, :expire_privileged_at)
-      .merge!(token: ENCRYPTOR.decrypt(attributes[:token]))
-      .with_indifferent_access
+  attr_readonly :token, :name, :uid, :kind, :cluster_id
 
-    new(params)
+  before_save :downcase_name
+
+  belongs_to :tokenable, -> { readonly }, polymorphic: true
```

Comment:
I think you can do `readonly: true` on the association instead.

Reply:
Nope. It's `-> { readonly }` now.

Reply:
Ah! Good to know.